server/src/services/semanticHighlight/onSemanticTokensFull.ts

```diff
@@ -0,0 +1,154 @@
+import { NumberHighlighter } from "./highlighters/NumberHighlighter";
+import { StringHighlighter } from "./highlighters/StringHighlighter";
+import { FunctionDefinitionHighlighter } from "./highlighters/FunctionDefinitionHighlighter";
+import { FunctionCallHighlighter } from "./highlighters/FunctionCallHighlighter";
+import { EventEmissionHighlighter } from "./highlighters/EventEmissionHighlighter";
+import { EventDefinitionHighlighter } from "./highlighters/EventDefinitionHighlighter";
+import { ContractDefinitionHighlighter } from "./highlighters/ContractDefinitionHighlighter";
+import { InterfaceDefinitionHighlighter } from "./highlighters/InterfaceDefinitionHighlighter";
+import { StructDefinitionHighlighter } from "./highlighters/StructDefinitionHighlighter";
+import { HighlightVisitor } from "./HighlightVisitor";
+
+const emptyResponse: SemanticTokens = { data: [] };
+
+export function onSemanticTokensFull(serverState: ServerState) {
+  return (params: SemanticTokensParams): SemanticTokens => {
+    const { telemetry, logger, solcVersions } = serverState;
+
+    return (
+      telemetry.trackTimingSync("onSemanticTokensFull", (transaction) => {
+        const { uri } = params.textDocument;
+
+        // Find the file in the documents collection
+        const document = serverState.documents.get(uri);
+
+        if (document === undefined) {
+          logger.error("document not found in collection");
+          return {
+            status: "internal_error",
+            result: emptyResponse,
+          };
+        }
+
+        const text = document.getText();
+
+        // Get the document's solidity version
+        let span = transaction.startChild({ op: "solidity-analyzer" });
+        const { versionPragmas } = analyze(text);
+        span.finish();
+
+        versionPragmas.push("<= 0.8.19"); // latest supported by slang
+
+        const solcVersion =
+          semver.maxSatisfying(solcVersions, versionPragmas.join(" ")) ||
+          _.last(solcVersions);
+
+        try {
+          // Parse using slang
+          span = transaction.startChild({ op: "slang-parsing" });
+          const language = new Language(solcVersion!);
+
+          const parseOutput = language.parse(
+            ProductionKind.SourceUnit,
+            document.getText()
+          );
+
+          const parseTree = parseOutput.parseTree;
+          span.finish();
+
+          if (parseTree === null) {
+            logger.error("Slang parsing error");
+            const strings = parseOutput.errors.map((e: any) =>
+              e.toErrorReport(uri, text, false)
+            );
+            logger.error(strings.join(""));
+
+            return {
+              status: "internal_error",
+              result: emptyResponse,
+            };
+          }
+
+          // Register visitors
+          const builder = new SemanticTokensBuilder(document);
+
+          const visitors = [
+            new CustomTypeHighlighter(document, builder),
```

Comment:
I wonder if we need to consider these additional types:

- Enums
- User defined value types.
- Errors.
- Libraries.
- etc...

Reply:
Definitely! I don't know how I missed those. Maybe because I covered them on document symbols and then forgot about highlighting

Reply:
Added support for these